File: Sources/structy/05_binary_tree/AllTreePaths.swift
//
//  AllTreePaths.swift
//  MyCLI
//
//  Created by new on 3/29/26.
//

func allTreePaths(_ root: TreeNode<String>?) -> [[String]] {

    func all_tree_paths(_ node: TreeNode<String>?) -> [[String]] {
        guard let node else {
            return [[]]
        }

        if node.left == nil && node.right == nil {
            return [[node.val]]
        }

        var res: [[String]] = []

        if let left = node.left {
            for var path in all_tree_paths(left) {
                path.append(node.val)
                res.append(path)
            }
        }

        if let right = node.right {
            for var path in all_tree_paths(right) {
                path.append(node.val)
                res.append(path)
            }
        }

        return res
    }

    var res = all_tree_paths(root)
    for i in res.indices {
        res[i].reverse()
    }

    return res
}

File: Sources/structyC/05_binary_tree/all_tree_paths.cpp
//
//  all_three_paths.h
//  MyCLI
//
//  Created by new on 3/29/26.
//

#include <vector>
#include <string>

class Node {
public:
    std::string val;
    Node* left;
    Node* right;

    Node(std::string initialVal) {
        val = initialVal;
        left = nullptr;
        right = nullptr;
    }
};

void all_tree_paths(Node *node, std::vector<std::string> curr, std::vector<std::vector<std::string>>* res) {
    if (node == nullptr) {
        return;
    }

    curr.push_back(node->val);

    if (node->left == nullptr && node->right == nullptr) {
        res->push_back(curr);
    }

    if (node->left != nullptr) {
        all_tree_paths(node->left, curr, res);
    }

    if (node->right != nullptr) {
        all_tree_paths(node->right, curr, res);
    }
}

std::vector<std::vector<std::string>> allTreePaths(Node* root) {
    std::vector<std::vector<std::string>> res{};
    all_tree_paths(root, std::vector<std::string>{}, &res);

    return res;
}

